### PR TITLE
[GEP-26] Add Support for Workload Identity Signing Key Rotation

### PR DESCRIPTION
**Pull request by vpnachev**
**How to categorize this PR?**
/area security ipcei
/kind enhancement
/ipcei workload-identity

**What this PR does / why we need it**:
Add Support for Workload Identity Signing Key Rotation

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
Part of https://github.com/gardener/gardener/issues/9586
/cc @dimityrmirchev 

**Release note**:

```other operator

```

### CODE REVIEW

File: pkg/component/gardener/apiserver/apiserver.go [hunk 1]
```diff
@@ -47,6 +47,8 @@ type Interface interface {
 	apiserver.Interface
 	// GetValues returns the current configuration values of the deployer.
 	GetValues() Values
+	// SetWorkloadIdentityKeyRotationPhase sets the current workload identity key rotation phase.
+	SetWorkloadIdentityKeyRotationPhase(gardencorev1beta1.CredentialsRotationPhase)
```
Comment (rfranzke) [inline pkg/component/gardener/apiserver/apiserver.go:51]: Nit: Not directly related to this PR, but could you move https://github.com/gardener/gardener/blob/129a182213ee1bbc27f857c78483841780facaba/pkg/component/gardener/apiserver/deployment.go#L103 into https://github.com/gardener/gardener/blob/129a182213ee1bbc27f857c78483841780facaba/pkg/component/gardener/apiserver/deployment.go#L172? When browsing the file, I was confused why this CLI flag is not part of this function. Thank you!

**Review by rfranzke — CHANGES_REQUESTED** [review]
Nice PR, thank you! Please augment https://github.com/gardener/gardener/blob/master/docs/concepts/operator.md#credentials-rotation.

Question: After the rotation of this key is prepared (i.e., a new key got created and GAPI got rolled out), shouldn't we issue new tokens for all `WorkloadIdentity`s in the system? Otherwise, how do we prevent that the completion is triggered too fast/early?

**Comment by vpnachev** [comment]
> Question: After the rotation of this key is prepared (i.e., a new key got created and GAPI got rolled out), shouldn't we issue new tokens for all `WorkloadIdentity`s in the system? Otherwise, how do we prevent that the completion is triggered too fast/early?

Good question, I was expecting the human operator to not trigger completion so fast, but as there is no prevention mechanism this could easily happen. I will think about this, but maybe a solution similar to [service account signing key rotation](https://github.com/gardener/gardener/blob/master/docs/extensions/garden-api-access.md#renewing-all-garden-access-secrets) will be the way forward, what do you think?

**Comment by rfranzke** [comment]
> > Question: After the rotation of this key is prepared (i.e., a new key got created and GAPI got rolled out), shouldn't we issue new tokens for all `WorkloadIdentity`s in the system? Otherwise, how do we prevent that the completion is triggered too fast/early?
> 
> Good question, I was expecting the human operator to not trigger completion so fast, but as there is no prevention mechanism this could easily happen. I will think about this, but maybe a solution similar to [service account signing key rotation](https://github.com/gardener/gardener/blob/master/docs/extensions/garden-api-access.md#renewing-all-garden-access-secrets) will be the way forward, what do you think?

Yes, this makes sense. I guess we don't need such operation annotation in the setup w/ `gardener-operator` since we can just trigger this in the reconciliation flow. We do this here for the `ServiceAccount` secrets: https://github.com/gardener/gardener/blob/3a13ee1aa75ee7e6c500aaec69326cd0c8640037/pkg/operator/controller/garden/garden/reconciler_reconcile.go#L319-L329 and https://github.com/gardener/gardener/blob/3a13ee1aa75ee7e6c500aaec69326cd0c8640037/pkg/operator/controller/garden/garden/reconciler_reconcile.go#L357-L380
I guess we should add a similar task for the `WorkloadIdentity` secrets

**Comment by dimityrmirchev** [comment]
Thanks! Apart from the https://github.com/gardener/gardener/pull/10313#pullrequestreview-2232788377 discussion other looks good to me so feel free move the PR forward once the discussion is resolved

**Comment by vpnachev** [comment]
Workload identity tokens in all seeds are now renewed with [297a091](https://github.com/gardener/gardener/pull/10313/commits/297a0918927d8ae2aafd6d5dc6e86d1ceea05d45)

File: pkg/utils/gardener/tokenrequest/secrets.go [hunk 2]
```diff
@@ -61,15 +68,29 @@ func GenerateGenericTokenKubeconfig(ctx context.Context, secretsManager secretsm
...
 // This will make the token-requestor controller in gardener-resource-manager/gardenlet issue new tokens immediately.
 func RenewAccessSecrets(ctx context.Context, c client.Client, opts ...client.ListOption) error {
+	// Apply to secrets labeled with resources.gardener.cloud/purpose=token-requestor.
+	return renewTokenInSecrets(ctx, c, resourcesv1alpha1.ServiceAccountTokenRenewTimestamp, *tokenRequestorRequirement, opts...)
+}
+
+// RenewWorkloadIdentityTokens drops the workloadidentity.security.gardener.cloud/token-renew-timestamp annotation
+// from all token secrets selected by the given list options. This will make the token-requestor-workload-identity
+// controller in gardenlet to issue new tokens immediately.
+func RenewWorkloadIdentityTokens(ctx context.Context, c client.Client, opts ...client.ListOption) error {
```
Comment (rfranzke) [inline pkg/utils/gardener/tokenrequest/secrets.go:78]: Does it make sense to write a small unit test for it?

**Review by rfranzke — COMMENTED** [review]
Very nice! Please augment https://github.com/gardener/gardener/blob/master/docs/concepts/operator.md#credentials-rotation

**Comment by rfranzke** [comment]
> Thanks! Apart from the [#10313 (review)](https://github.com/gardener/gardener/pull/10313#pullrequestreview-2232788377) discussion other looks good to me so feel free move the PR forward once the discussion is resolved

@dimityrmirchev Do you plan to have a second look as well?

**Comment by dimityrmirchev** [comment]
> @dimityrmirchev Do you plan to have a second look as well?

Yes, I will take a look in the afternoon

**Review by dimityrmirchev — APPROVED** [review]
/lgtm

File: pkg/utils/gardener/tokenrequest/secrets.go (hunk 2 above)
Comment (vpnachev) [reply pkg/utils/gardener/tokenrequest/secrets.go:78]: Sure, done with 37d4fd421e98bfde22eda2cc8799ff5429625ce3

File: docs/concepts/operator.md [hunk 3]
```diff
@@ -658,6 +659,9 @@ Similarly, when the CA certificate rotation is in `Preparing` phase, then `garde
 This causes `gardenlet` to request a new client certificate for its garden cluster kubeconfig, which is now signed with the new client CA, and which also contains the new CA bundle for the server certificate verification.
 Read more about it [here](gardenlet.md#rotate-certificates-using-bootstrap-kubeconfig).
 
+Also, when the `WorkloadIdentity` token signing key is in `Preparing` phase, then `gardener-operator` annotates all `Seed`s with `gardener.cloud/operation=renew-workload-identity-tokens`.
```
Comment (rfranzke) [inline docs/concepts/operator.md]:
```suggestion
Also, when the `WorkloadIdentity` token signing key rotation is in `Preparing` phase, then `gardener-operator` annotates all `Seed`s with `gardener.cloud/operation=renew-workload-identity-tokens`.
```

**Pull request merged**